controllers/external/util.go

```diff
@@ -169,6 +179,23 @@ func GenerateTemplate(in *GenerateTemplateInput) (*unstructured.Unstructured, er
 	return to, nil
 }
 
+func SetCloneRefAnnotation(ref corev1.ObjectReference, obj *unstructured.Unstructured) error {
```

Comment:
I'm not sure we should be adding additional use of `corev1ObjectReference` anywhere, especially if we are stuffing it into an annotation. Would it make sense to either do some type of string concatenation of the api group, kind, and resource name instead? Or possibly use separate annotations for each?

Reply:
I thought about this (also just noticed that we need a godoc string on this exported function) and I agree that we're adding to the problem; we _should_ though be able to convert to our own type later on with conversion webhooks?

Reply:
I think one concern I have here is that if we are using corev1.ObjectReference here for comparison purposes, then we'll still have to do manipulation of the reference to do things like ignoring the api version when doing comparisons anyway, especially in the case of externally managed infrastructure providers.

It would also seem error prone to expect all providers to implement a conversion webhook to unmarshal a corev1.ObjectReference (where the type may change in the future) from an annotation in order to update the api version associated with that reference.

Reply:
Makes sense, do we want to introduce our own clusterv1.Reference type today?

Reply:
Any particular reason to prefer a type than a concatenation convention (or multiple annotations)? I worry that trying to shoehorn a type object into a string field is going to lead to pain later on since there is no api machinery involved in accessing or conversion.

Reply:
Agree if it was labels, I've always assumed labels are more for controllers to use for internal things, having a typed json struct is very common and I was kinda expecting to move to a typed contract later, for example have external references to have a `spec.ClonedFrom` reference that points to the actual object. Multiple annotations might work, although we'll need to provide utilities to ease how to read these, having concatenation seems we're creating a new standard when json already solves this 🤷

Reply:
@detiber @vincepri If there are plans to have` spec.ClonedFrom` field and this PR solves the issue temporarily, leaving it as is would make sense IMO as json unmarshalling makes things easier. Or concatenation approach would be my second choice because creating a narrower object ref type would affect many pieces, that may need a bit more discussion.

Reply:
Can we find any existing precedent for storing json-encoded typed data in annotations in the k8s community (outside of our use within conversion webhooks)?

I'm purposefully wanting to ignore our use with conversion webhooks because it is not something that is exposed as part of the contract between different components and we strip the information out prior to storing the resource, which prevents any issues with marshalling based on different versions between local client and the stored version.

Reply:
[TolerationsAnnotationKey](https://github.com/kubernetes/kubernetes/blob/36981002246682ed7dc4de54ccc2a96c1a0cbbdb/pkg/apis/core/annotation_key_constants.go#L34) 
  There is an example of type `Toleration` being marshalled and added as annotation.

Thanks for the explanation @detiber , then I will go with concatenation approach where Group, Kind, and Name is concatenated for the annotation.